### PR TITLE
IMPRO-1209 Callable plugins 

### PR DESCRIPTION
**Pull request by neilCrosswaite**
Makes plugins callable 
Added an Abstract Base Class

plugins, all classes with a process method, inherit from that base class.

RegridLandSea does not at the moment as it makes the tets fail.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

File: lib/improver/__init__.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                Positional arguments.
+            **kwargs:
+                Keyword arguments.
+        Returns:
+            Output of self.__call__()
+        """
+        return self.__call__(*args, **kwargs)
+
+    @abstractmethod
+    def __call__(self):
```
Comment (TomekTrzeciak) [inline lib/improver/__init__.py]: I wonder if we shouldn't do it the other way around - make `process` into abstract method (effectively an interface that each plugin has to implement) and call it in `__call__` dunder. I think this is quite common implementation pattern (also, the diff for this change will be smaller). @LaurenceBeard, any opinion on this?

Comment (LaurenceBeard) [reply lib/improver/__init__.py]: ^ Yes, this is cleaner, would maintain backwards compatibility also (eg. notebooks).

Comment (neilCrosswaite) [reply lib/improver/__init__.py]: I have now changed in as requested.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -38,6 +38,7 @@ below:
 * Benjamin Ayliffe (Met Office, UK)
 * Mark Baker (Met Office, UK)
 * Laurence Beard (Met Office, UK)
+Neil Crosswaite (Met Office, UK)
```
Comment (PaulAbernethy) [inline CONTRIBUTING.md]: This needs an * to make it a bullet point

File: lib/improver/__init__.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+from abc import ABC, abstractmethod
+
+
+class BasePlugin(ABC):
+    """An abstract class for IMPROVER plugins.
+    Subclasses must be callable. We preserve the process
+    method by redirecting to __call__.
+    """
+    def __call__(self, *args, **kwargs):
+        """ Makes subcasses callable to use process
```
Comment (PaulAbernethy) [inline lib/improver/__init__.py]: typo: subcasses -> subclasses

Comment (neilCrosswaite) [reply lib/improver/__init__.py]: Fixed it.

File: CONTRIBUTING.md (hunk 2 above)
Comment (neilCrosswaite) [reply CONTRIBUTING.md]: Fixed it

**Comment by arh89** [comment]
@TomekTrzeciak, @LaurenceBeard - it's a shame that this has been left to have conflicts. imo, this was low hanging fruit and the longer we leave this, the more likely there will be (even more) conflicts.

@neilCrosswaite, once you have time (check with whoever the blue team scrum master is, @bayliffe, @MoseleyS?), could you resolve the conflicts please? Once this is done, I will review and merge. One review is sufficient, given the number of discussions around this topic.

(Sorry for being so slow to jump on this)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=3830094&bid=13382423)

File: lib/improver/cli/nowcast_optical_flow.py [hunk 4]
```diff
@@ -35,7 +35,6 @@
 import os
 
 import iris
-import numpy as np
```
Comment (arh89) [inline lib/improver/cli/nowcast_optical_flow.py]: How come this has been removed?

File: lib/improver/ensemble_copula_coupling/ensemble_copula_coupling_utilities.py [hunk 5]
```diff
@@ -34,7 +34,6 @@
 
 """
 import copy
-import random
```
Comment (arh89) [inline lib/improver/ensemble_copula_coupling/ensemble_copula_coupling_utilities.py]: Shouldn't have been removed?

File: lib/improver/utilities/spatial.py [hunk 6]
```diff
@@ -38,9 +38,9 @@
 import scipy.ndimage
 from iris.coords import CellMethod
 from iris.cube import Cube, CubeList
-from iris.exceptions import CoordinateNotFoundError
```
Comment (arh89) [inline lib/improver/utilities/spatial.py]: Shouldn't have been removed?

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=3830094&bid=13382423)

**Comment by arh89** [comment]
(Neil says the imports are unnecessary - he's going to raise those in a separate PR)

**Comment by arh89** [comment]
Classes which don't inherit from BasePlugin
```
[~/improver/lib/improver]$ grep --exclude-dir=tests -rP '^class.*:$' * | grep -v BasePlugin
argparser.py:class ArgParser(ArgumentParser):
blending/weighted_blend.py:class PercentileBlendingAggregator:
blending/weights.py:class WeightsUtilities:
cli/__init__.py:class HelpForNapoleonDocstring(HelpForAutodetectedDocstring):
cli/__init__.py:class DocutilizeClizeHelp(ClizeHelp):
cli/__init__.py:class ObjectAsStr(str):
ensemble_calibration/ensemble_calibration.py:class ContinuousRankedProbabilityScoreMinimisers:
lapse_rate.py:class SaveNeighbourhood(object):
nbhood/circular_kernel.py:class CircularNeighbourhood(object):
nbhood/circular_kernel.py:class GeneratePercentilesFromACircularNeighbourhood(object):
nbhood/nbhood.py:class GeneratePercentilesFromANeighbourhood(BaseNeighbourhoodProcessing):
nbhood/nbhood.py:class NeighbourhoodProcessing(BaseNeighbourhoodProcessing):
nbhood/square_kernel.py:class SquareNeighbourhood(object):
psychrometric_calculations/psychrometric_calculations.py:class Utilities(object):
utilities/spatial.py:class OccurrenceWithinVicinity:
utilities/spatial.py:class RegridLandSea:
utilities/warnings_handler.py:class ManageWarnings(object):
wind_calculations/wind_downscaling.py:class RoughnessCorrectionUtilities(object):
```

**Comment by arh89** [comment]
CLI tests pass aside from these 3:
```
improver-nowcast-optical-flow/05-metadata.bats
improver-nowcast-optical-flow/07-basic_no_orographic_enhancement.bats
improver-nowcast-optical-flow/08-remasked.bats
```
which I suspect are unrelated to these changes (possibly to do with using a different version of numpy, picked up from ppdev) - testing `master` now to confirm no difference between the two...

**Pull request merged**